### DISCUSSION TITLE
n8n Integration

### DISCUSSION THREAD

**Discussion by aibucketmail** [langfuse/langfuse#4397] [Ideas]
### Describe the feature or potential improvement

Able to track n8n workflow 

> [!NOTE]  
> Official support for tracing n8n workflows relies on the addition of this feature as a native option in n8n. This would include tracking spans for each node in a workflow and not only logging llm calls to langfuse.
>
> Given popular demand, @rorubyy has created a community node ([rorubyy/n8n-nodes-openai-langfuse](https://github.com/rorubyy/n8n-nodes-openai-langfuse)) to trace OpenAI and OpenAI-compatible models in n8n.

**Comment by RepairYourTech** [comment]
I 100% second a need for this, I already use langfuse woth flowise in my k8s and want to be able to monitor n8n as well, since I use n8n to compliment flowise

**Comment by marcklingen** [comment]
I would love to support this integration. As n8n seems to be based on LangChain, the existing [LangChain integration](https://langfuse.com/docs/integrations/langchain/tracing) could be reused (similar to the Langfuse integrations with Langflow and Flowise).

If you want to contribute this to n8n or find someone who is interested on the n8n team to build this, I am happy to help. As n8n is a platform product, we cannot build an external integration as this needs to be added to the n8n project itself.

**Comment by marcklingen** [comment]
Fyi, there is an ongoing thread for this here: https://community.n8n.io/t/swap-smith-langchain-for-langfuse/47748/9

It'd be awesome if you could contribute this to n8n core as this integration might be relevant for many other teams as well.

**Comment by pnajdecki** [comment]
The n8n framework is a seasoned solution that is ideal for rapid prototyping (and more) using AI. Support from LangFuse and integration with n8n will significantly increase the quality and efficiency of such solutions, as well as reduce their development time and the time to solve potential problems.

**Comment by RepairYourTech** [comment]
Bump

**Comment by marcklingen** [comment]
@mayphilc Nothing we can do here unfortunately, this needs to be added on the n8n side

**Comment by oussamadouhou** [comment]
There is a feature request you can vote for:

https://community.n8n.io/t/capturing-n8n-flows-with-observability-evaluations-platforms-like-langfuse-comet-opik-etc/75764

**Comment by traderpedroso** [comment]
bump

**Comment by simple10** [comment]
Bump. FYI, proxying n8n AI calls through LiteLLM with Langfuse works really well until n8n has native observability fleshed out.

**Reply by pnajdecki** [reply]
That is interesting. Would you please provide more insights (or links, etc. ) on this approach?

**Reply by marcklingen** [reply]
More on this integration here: https://oss-llmops-stack.com (mutual website maintained by litellm and langfuse teams)

tldr: you can set up litellm as a proxy for all openai calls and forward them to langfuse via the callbacks. This does not replace tracing as you do not have the full visibility into the application but at least you have a log of llm calls.

**Reply by RepairYourTech** [reply]
No kidding wow, I assume this works with all openai compatible providers like groq and deepseek? Even though this won't give you real observability does it at least give you a mostly accurate token count?

**Reply by marcklingen** [reply]
Yes, this gives you logs and token counts. You can thereby also use 100+ models via the n8n openai connector as litellm will translate models to the openai api spec

**Reply by pnajdecki** [reply]
Thank you, In the meantime, I managed to run this successfully. It works like a charm (i used promptfoo + LiteLLM + Langfuse)

**Reply by RepairYourTech** [reply]
Can I just sneak litellm in my existing stack? I already have n8n and flowise and a few others all set up, including langfuse. I dont really want to redo langfuse that was a nightmare

**Comment by tan-yong-sheng** [comment]
Perhaps can just add a litellm node on n8n also instead of just using openai node? Because if using litellm proxy, you could set custom metadata such as 'user_id', etc to pass to langfuse for better observability.

Something like this, as shown in the litellm docs: https://docs.litellm.ai/docs/observability/langfuse_integration, which is no way (or I haven't found it yet) to set so if using openai node:
```
curl --location --request POST 'http://0.0.0.0:4000/chat/completions' \
    --header 'Content-Type: application/json' \
    --header 'Authorization: Bearer sk-1234' \
    --header 'langfuse_trace_id: trace-id2' \
    --header 'langfuse_trace_user_id: user-id2' \
    --header 'langfuse_trace_metadata: {"key":"value"}' \
    --data '{
    "model": "gpt-3.5-turbo",
    "messages": [
        {
        "role": "user",
        "content": "what llm are you"
        }
    ]
}'
```

**Reply by marcklingen** [reply]
Thanks for sharing! Are you open to contribute a documentation page on this? It seems like you have figure this out end-to-end while many other folks are struggling to set this up correctly

**Reply by tan-yong-sheng** [reply]
> Perhaps can just add a LiteLLM node on n8n also instead of just using openai node?

Just to clarify, my suggestion is to let n8n side to add LiteLLM node (and langfuse need not to update anything). 

> What this LiteLLM node in n8n could be used for?

Only those who self-hosted LiteLLM proxy website could be able to use it like openai node in n8n (and they will need to change API key and base_url to LiteLLM proxy's website address as usual).

But, then the disadvantage of this implementation is that the user might be forced to be tied with LiteLLM ecosystem in future, just to pass custom metadata to langfuse as LiteLLM proxy website will act as intermediary.

Also, adding 1 more infrastructure like LiteLLM proxy might be troublesome for someone just requires simple logging functionalities between n8n and langfuse.

> Why not using existing openai node as litellm is also openai-compatible.

As explained above, while we can use existing openai node, but we can't pass in custom metadata like 'user_id' to langfuse (and perhaps other logging capabilities) if doing so.

To be honest, I haven't yet fully explored the logging capabilities of using LiteLLM proxy website, and this is the docs I found for logging capabilities between LiteLLM proxy and langfuse: https://docs.litellm.ai/docs/observability/langfuse_integration

But just in case you just needs to simple guide on setting up LiteLLM proxy, I wrote it before: https://www.tanyongsheng.com/note/running-litellm-and-openwebui-on-windows-localhost-a-comprehensive-guide/ ... But I haven't yet added langfuse integration there yet.

------

Hopefully it's clear to make sure we're in sync. Feel free to discuss, thanks a lot 🙏

Ya, @marcklingen I am happy to help to document if you guys are interested with this proposed solution. Just let me know what criteria need to be included in the documentation.

**Reply by marcklingen** [reply]
Got it, seems like a good extension of n8n would be to add support for additional request metadata / headers which could be picked up by the litellm/langfuse integration

**Comment by raphael2692** [comment]
Bump

**Comment by DrLuthor** [comment]
Bump

**Comment by marcklingen** [comment]
Great to see the interest on this here. Unfortunately this very much depends on open tracing/instrumentation capabilities in n8n which do not exist (need to make this happen in n8n in order for other tools to integrate for tracing/evals such as langfuse).

In the meantime the litellm proxy route seems to be the best one, although it only provides llm logs and does not include detailed agent traces.

**Reply by RepairYourTech** [reply]
Is it a matter of making n8n OTel compatible? I pretty much just stopped using n8n altogether and started using googles ai adk instead because I can't get any metrics from my workflows which is pretty much useless in production so as badass as n8n is unfortunately I have to put it on the shelf like a toy I want to play with but I know if I do ill just be held back.

**Reply by marcklingen** [reply]
Yes, I think the best integration path would be via adding OTel support to n8n

**Reply by RepairYourTech** [reply]
Check this out, might be a temporary workaround?

https://community.n8n.io/t/n8n-successfully-instrumented-with-opentelemetry/78468

At least until the team at n8n get their heads out of the sand?

**Comment by RepairYourTech** [comment]
Check this out, looks like a slightly better workaround than the other one 

https://community.n8n.io/t/n8n-successfully-instrumented-with-opentelemetry/78468

**Comment by Tonvenio** [comment]
Bump

**Comment by deichrenner** [comment]
Bump

**Comment by biliboss** [comment]
I've created a workflow that receives the execution id and use REST API to send the trace. Let me know if anyone is interested in this code. It load execution data (also sub workflows) and map to API.

**Reply by RepairYourTech** [reply]
Until n8n adds support this is just sad

**Reply by oscarfesa** [reply]
Totally interested!

**Reply by ericyangliu** [reply]
I'm interested! Do you mind sharing the code?

**Reply by GTHell** [reply]
3 people already let you know. where is the code?

**Comment by ericyangliu** [comment]
Bump

**Comment by marcklingen** [comment]
We’ve just released a node for n8n. Currently it supports Langfuse prompt management. It is pending verification which is required for using it in n8n cloud. If you self host n8n, you can install the node via the community node settings.

https://langfuse.com/changelog/2025-07-06-n8n-node-for-langfuse-prompts

No update on tracing integration with n8n yet, as this depends on platform level support within n8n.

**Reply by DavideDelbianco** [reply]
good luck receiving a response from n8n team, they have been ignoring other very popular threads for 2 years  now

**Comment by RepairYourTech** [comment]
This is why i refuse tonuse or promote n8n

**Comment by hudsonbrendon** [comment]
bump

**Comment by estebanrestoyagicap** [comment]
bump

**Comment by cymulatereouven** [comment]
bump

**Comment by maudelv** [comment]
bump

**Comment by mjmendo** [comment]
bump

**Comment by LaloLalo1999** [comment]
bump, would love tracing integration/node

**Comment by rodrigocpaiva** [comment]
Bump

**Comment by juanmartitegui** [comment]
bump

**Comment by user823453** [comment]
needed

**Comment by RepairYourTech** [comment]
They are not going to do this. I saybthis because its been a year of people requesting it and ive even reached out directly and they told me they have so much going on MAYBE  they will look into it in 2026 at some point. 

I know the majority of people use the community edition but imagine wanting n8n for professional use and paying the insane price they charge to self host it and not only being artificially limited to a set amount of executions but also the team ignores such demand and interest for a feature like this. 

Make no mistake the people at n8n are greedy pieces of crap. They care nithing about the userbase and only about being able to charge you $800/mo for artificially limited software. 

Its time for a replacement to n8n i think we need to work on an open source alternative and let n8n feel the sting of disinterest

**Comment by rorubyy** [comment]
Hi there! 👋 Just wanted to share that we built a custom n8n community node that integrates OpenAI and Langfuse together 🎯

**n8n-nodes-openai-langfuse**
Github: [https://github.com/rorubyy/n8n-nodes-openai-langfuse](https://github.com/rorubyy/n8n-nodes-openai-langfuse)
npm: [https://www.npmjs.com/package/n8n-nodes-openai-langfuse](https://www.npmjs.com/package/n8n-nodes-openai-langfuse) 

It helps you easily trace and analyze LLM usage inside n8n using Langfuse, including session metadata support.

Hope it helps!

**Reply by marcklingen** [reply]
Thank you for your effort in sharing this! While we wait for official OpenTelemetry support for n8n, this appears to be a great workaround for many people in this thread.

I have added it to the integration documentation page to make it easier to discover: https://github.com/langfuse/langfuse-docs/commit/94f7bc0942981332905557dfcc0559abe668cdf2

**Reply by DavideDelbianco** [reply]
Are you hinting that we might be able to get full workflow execution nodes traced in langfuse by OTEL integration between n8n and langfuse?
Is this something they are working on, or just something you hope for the future?

**Reply by marcklingen** [reply]
I hope for OTel-tracing support in n8n, as it makes sense for both AI and non-AI applications when using n8n at larger scale.

I am not a maintainer or contributor to n8n (we are pretty busy with maintaining/improving langfuse), so I do not know whether this is on the n8n roadmap.

**Reply by rorubyy** [reply]
I’ve just released a new node: n8n-nodes-ai-agent-langfuse
The earlier [OpenAI-based version](https://github.com/rorubyy/n8n-nodes-openai-langfuse) only traced the LLM, missing tool calls and RAG steps.
This update shifts tracing to the chain level, giving full visibility into tools, memory, and reasoning.
👉 https://github.com/rorubyy/n8n-nodes-ai-agent-langfuse

<img width="434" height="252" alt="agent-example" src="https://github.com/user-attachments/assets/023c4532-d99d-4d25-8e19-7c85c4c9192c" />

**Reply by alimohammad-123** [reply]
Hey does this only work with OpenAI? or can I use google gemini models as well?

**Comment by marcklingen** [comment]
New work by @rwb-truelime to patch self-hosted n8n to add Langfuse tracing

Original post on n8n community forum: https://community.n8n.io/t/capturing-n8n-flows-with-observability-evaluations-platforms-like-langfuse-comet-opik-etc/75764/13

Proof of concept: https://github.com/rwb-truelime/n8n-langfuse

**Reply by rwb-truelime** [reply]
Hi @marcklingen, just to let you and others know. I am also working on a Python shipper variant that can ship in batches instead of realtime: https://github.com/rwb-truelime/n8n-langfuse-shipper. Cheers

**Comment by sesigl** [comment]
Hey @marcklingen, did you consider provided another langfuse node from your side? Although not ideal, we could make use of n8n agent provided `intermediateSteps` ([n8n doc](https://docs.n8n.io/integrations/builtin/cluster-nodes/root-nodes/n8n-nodes-langchain.agent/tools-agent/#return-intermediate-steps)) . They contain a lot of information relevant to create proper tracing logs.

Thanks !

**Comment by genelim** [comment]
Bump

**Comment by GuiFlam** [comment]
bump

**Comment by edgBR** [comment]
bump!

**Comment by mykola-mmm** [comment]
+++

**Comment by rcwmt** [comment]
Would be amazing

**Comment by Aiman-E** [comment]
Bump

**Comment by ghost** [comment]
This would be great!

**Comment by avatsaev** [comment]
100% we need this

**Comment by RepairYourTech** [comment]
There no point bumping this. Theres nothing langfuse can do about it. You have to bother the n8n devs. Ive been pestering them for a year now to do it and the only response i ever got was "we have too much on the horizon right now but we might be able to look into this late 2026-2027"

So its alost cause they dont give a rats ass. 

Let n8n die out they only care about enterprise customers and there are plenty of better ways to build agents now, just tell any coding agent in vs code to build you pydanticai agents for whatever you want with a streamlit chat ui

**Comment by Lotte-Verheyden** [comment]
A user recently mentioned using OpenRouter in N8N, and then using the [OpenRouter Broadcast feature](https://langfuse.com/integrations/gateways/openrouter#broadcast) to get the traces into Langfuse. Works very well, can be a good workaround for a lot of people in the meantime :)

**Reply by javierlabanca** [reply]
@Lotte-Verheyden Have you managed to send tracing metadata (like sessionId, userId, etc) to Openrouter broadcast to Langfuse? My n8n traces appear on Langfuse, but without metadata (which somehow gets lost on the way), which makes it almost impossible to analyze.

**Reply by sdandois** [reply]
We were able to include metadata using the Open AI model node connected to Open Router by changing base url in credentials and adding metadata object to the node settings.

**Reply by javierlabanca** [reply]
Thank you @sdandois. Your comment has been very helpful! One further question I have: the Openai chat model seems to be wrapping the metadata inside the metadata key, which I think is preventing Openrouter from reading the values. I'm attaching an image.

How were you able to solve that issue if you'd be so kind to tell me?
<img width="411" height="490" alt="metadata" src="https://github.com/user-attachments/assets/6015f2c5-d507-4bd2-b668-3b7a9f72a623" />

**Reply by javierlabanca** [reply]
To clarify: that image is the request n8n is sending to Openrouter.

**Reply by sdandois** [reply]
the metadata object is transformed to `metadata_` prefix. It just works.

<img width="711" height="122" alt="Screenshot 2026-04-29 at 15 35 57" src="https://github.com/user-attachments/assets/0520b16d-cd12-46c2-917a-eae21ce0d6d4" />

_image is from langfuse tracing_

I have a question of my own, how were you able to inspect the raw HTTP request behind the model node? 

I've been wandering that for a while.

**Reply by javierlabanca** [reply]
Thank you, sdandois! I see it now. It's storing the metadata with a prefix, which is kind of useful. Nevertheless, it's not populating the Sessions tab in Langfuse, which tracks the conversations as per user session, which makes it much more useful to follow and annotate! I've found that, in order for Langfuse to properly track sessions in a new tab, which each trace contained in the session, you have to include the session_id key outside the metadata object. Like this:

<img width="1041" height="543" alt="metadata2" src="https://github.com/user-attachments/assets/07d52307-1255-414a-a967-75b166d0d4a1" />

To inspect the raw HTTP request that n8n sends out into the world, I momentarily pointed the OpenAI chat model's base URL in the credentials to a webhook on this site https://webhook.site/.

**Comment by sdandois** [comment]
Please, take a look to this n8n plugin for OTEL observability with langwatch and also opik fork:
https://github.com/langwatch/n8n-observability

**Comment by apaul314** [comment]
+++

**Comment by devcaiosa** [comment]
+++

**Comment by lanmi42** [comment]
+1
